Add gradcheck for forward AD by default and functional API

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49529 Add more formulas
* **#49099 Add gradcheck for forward AD by default and functional API**
* #49098 Add formulas and basic tests

RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

albanD committed:
Add gradcheck for forward AD by default and functional API

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"




[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -201,6 +202,40 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
+                if res is None:
+                    jacobian[i][lin_idx].zero_()
+                else:
+                    jacobian[i][lin_idx].copy_(res.contiguous().view(-1))
```
nit: res.reshape(-1)

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -274,7 +275,7 @@ def vjp(func, inputs, v=None, create_graph=False, strict=False):
 
 
-def jvp(func, inputs, v=None, create_graph=False, strict=False):
+def jvp(func, inputs, v=None, create_graph=False, strict=False, fw_mode=False):
```
What's the plan for changing the default from False -> True?

albanD replied:
The plan is to wait until we have more formulas added. This will happen in a follow up of this stack.

zou3519 commented on test/test_autograd.py:
```diff
@@ -7197,7 +7219,8 @@ def func(root, b):
         gradcheck(func, [a, b], raise_exception=True)
         go = torch.randn(a.size(), device=device, requires_grad=True)
-        gradgradcheck(func, (a, b), (go,))
+        # TODO(albanD): re-enable forward tests
+        gradgradcheck(func, (a, b), (go,), check_forward=False)
```
Do in-place views not work at the moment?

albanD replied:
They do and are tested for.
This double backward test is the only one failing so I guess something with the test setting is not right. I will need to investigate further.

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -296,6 +297,9 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
             independent of it. If ``False``, we return a Tensor of zeros as the
             jvp for said inputs, which is the expected mathematical value.
             Defaults to ``False``.
+        fw_mode (bool, optional): If ``True``, forward mode AD will be used to
+            compute the jvp, otherwise, the backward of backward is used (sometimes
+            called the double backwards trick).
```
Nit: maybe we should say that the backward of backward is slow

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -342,18 +341,42 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
+            outputs_dual = func(*inputs_dual)
+            is_outputs_tuple, outputs_dual = _as_tuple(outputs_dual, "outputs of the user-provided function", "jvp")
+
+            outputs, grad_res = [], []
+            for o_dual in outputs_dual:
+                o_primal, o_dual = fwAD.unpack_dual(o_dual)
+                outputs.append(o_primal)
+                grad_res.append(o_dual)
```
nit:
```
outputs, grad_res = zip(*[fwAD.unpack_dual(o_dual) for o_dual in outputs_dual])
```

zou3519 reviewed (commented):
Some questions and minor comments but this looks pretty reasonable to me

albanD commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1834,11 +1834,11 @@ def __enter__(self):
         pass
 
-def _assertGradAndGradgradChecks(test_case, apply_fn, inputs):
+def _assertGradAndGradgradChecks(test_case, apply_fn, inputs, check_forward=True):
```
Thanks for the ref. The merge conflict should be fairly easy to handle indeed.

albanD commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5043,8 +5043,11 @@ def _do_test(self, test_case, module, input):
             # could probably unify check_jacobian above with this.
             params = tuple(x for x in module.parameters())
             num_inputs = len(input_tuple)
+            # These modules don't actually take the arguments as inputs so forward check will fail for these
```
Yes. The issue did not exist at that time, but I'll update this now. Thanks for pointing it out!

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -401,6 +445,27 @@ def not_reentrant_error(error_str=''):
             return not_reentrant_error(' (calculated using complex valued grad output)')
 
+        if check_forward:
+            try:
+                fw_analytical = get_analytical_jacobian_fw(fn, tupled_inputs, o)
+            except RuntimeError as e:
+                msg = str(e)
+                if "Trying to use forward prop with" in msg:
+                    warnings.warn("Failed to compute gradcheck using fw mode: {}".format(msg), stacklevel=stacklevel)
+                    continue
+                else:
+                    raise e
```
This error happens if the op has no forward formula in derivatives.yaml or the forward formula is defined as "not_implemented".

This check allows me to run the full test suite with forward AD checks on by default without needing to implement all the formulas.
This will be removed once more formulas are added and we can set the `check_forward=False` appropriately across the test suite.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -401,6 +445,27 @@ def not_reentrant_error(error_str=''):
+            except RuntimeError as e:
+                msg = str(e)
+                if "Trying to use forward prop with" in msg:
+                    warnings.warn("Failed to compute gradcheck using fw mode: {}".format(msg), stacklevel=stacklevel)
```
Yes.
Because warnings with default setting show only one line (and not the full stack), I think that pointing to the place where gradcheck is called is more important than this internal line within the gradcheck function.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -338,10 +380,12 @@ def fn(input):
         def fn(input):
             return _as_tuple(func(*input))[i]
 
+        numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
+
+        o = fn(tupled_inputs)
```
Yes, The forward is re-done here to prevent failures for some sparse tests when the numerical jacobian is done before the backward analytical one.
I am not 100% sure what is the root cause. But re-doing the forward seems like the right thing to do in any case.

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -401,6 +445,27 @@ def not_reentrant_error(error_str=''):
+                fw_analytical = get_analytical_jacobian_fw(fn, tupled_inputs, o)
+            except RuntimeError as e:
+                msg = str(e)
+                if "Trying to use forward prop with" in msg:
```
Yes this is ok for now until we get more general coverage for this.

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -278,6 +315,8 @@ def gradcheck(
         check_undefined_grad (bool, options): if True, check if undefined output grads
             are supported and treated as zeros, for ``Tensor`` outputs.
+        check_forward (bool, optional): if True, check the forward mode AD gradient
+        stacklevel (int, optional): the stack level to use for raising errors and warnings
```
This is only used for warnings, right? If so, nit: `warnings_stacklevel` may be a more descriptive name

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -296,6 +297,9 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
             independent of it. If ``False``, we return a Tensor of zeros as the
             jvp for said inputs, which is the expected mathematical value.
             Defaults to ``False``.
+        fw_mode (bool, optional): If ``True``, forward mode AD will be used to
+            compute the jvp, otherwise, the backward of backward is used (sometimes
+            called the double backwards trick) but it will be significantly slower.
```
nit: state default

zou3519 replied:
We should state the default (False)

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -234,7 +269,9 @@ def gradcheck(
     check_undefined_grad: bool = True,
-    check_grad_dtypes: bool = False
+    check_grad_dtypes: bool = False,
+    check_forward: bool = False,
```
nit: Every time I see `check_forward`, I get a little confused between the "forward pass" of a model and "forward-mode AD". Maybe it is better to make these flags more explicit? `check_forward_mode_ad`

zou3519 replied:
Everything I see "check_forward" I get confused:
- does it mean "check a forward pass?" or "forward-mode AD"?

Maybe we should clarify the name via `check_forward_mode_ad` or `check_jvp`

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5043,8 +5043,12 @@ def _do_test(self, test_case, module, input):
             num_inputs = len(input_tuple)
+            # These modules don't actually take the arguments as inputs so forward check will fail for these
+            # We will be able to re-enable this when https://github.com/pytorch/pytorch/issues/49171 is done
+            check_forward = len(params) == 0
```
nit: I had to stop here and think for a few seconds. The question that came to my mind was ,"when are there no parameters?" It would be nice to clarify in a comment that here we are checking forward mode AD for nn.Modules with no parameters like nn.Softmax, nn.Relu, ...

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5043,8 +5043,12 @@ def _do_test(self, test_case, module, input):
+            # We will be able to re-enable this when https://github.com/pytorch/pytorch/issues/49171 is done
+            check_forward = len(params) == 0
             _assertGradAndGradgradChecks(
-                test_case, lambda *args, **kw: test_case._forward(module, args[:num_inputs]), input_tuple + params)
+                test_case, lambda *args, **kw: test_case._forward(module, args[:num_inputs]), input_tuple + params,
+                check_forward=check_forward)
```
(no action required) Now, the very interesting thing here is that the grad and gradgrad checks only run if `check_gradgrad` is True! If `check_gradgrad` is False, then we don't run gradcheck, instead we go through `test_case.check_jacobian` above, which doesn't check forward-mode AD. `check_gradgrad=False` appears in a number of tests in test_nn, so we don't have coverage for those, but a lot of them weren't going to get the forward mode ad tests anyways because they're modules

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).



[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5044,6 +5044,10 @@ def fn_to_gradcheck(*inputs_and_params, **kwargs):
 
+        # These modules don't actually take the arguments as inputs so forward check will fail for these
+        # We will be able to re-enable this when https://github.com/pytorch/pytorch/issues/49171 is done
+        check_forward = len(params) == 0
```
We should be more descriptive here and name it "check_forward_mode_ad". Otherwise, there is the ambiguity if we are talking about forward mode AD or the forward pass of the module

albanD replied:
That's a very good point. I forgot that there was anything else but forward mode AD :D

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -274,7 +275,7 @@ def vjp(func, inputs, v=None, create_graph=False, strict=False):
 
 
-def jvp(func, inputs, v=None, create_graph=False, strict=False):
+def jvp(func, inputs, v=None, create_graph=False, strict=False, fw_mode=False):
```
The flag is called `forward_mode` in gradcheck -- we should probably be consistent? Also "fw" might not be the best abbreviation, if we wanted to abbreviate it "fwd" should be clearer

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -342,18 +341,37 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
-    grad_outputs = tuple(torch.zeros_like(out, requires_grad=True) for out in outputs)
+    if fw_mode:
+        if v is None:
+            v = (torch.ones_like(inputs[0]),)
```
nit: assert (or write a comment) that `inputs` is expected to have exactly 1 element. It took me a while to figure out that invariant

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -342,18 +341,37 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
+
+        grad_res = _autograd_grad(grad_inputs, grad_outputs, v, create_graph=create_graph)
 
     jvp = _fill_in_zeros(grad_res, outputs, strict, create_graph, "back_trick")
```
By "back_trick", do we mean the "double backward trick"? This is not necessarily the double backward trick anymore, especially if we use forward-mode AD, right?

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -401,6 +445,29 @@ def not_reentrant_error(error_str=''):
+                # Test was aborted while computing the jacobian
+                continue
+
+            for j, (a, n) in enumerate(zip(fw_analytical, numerical)):
+                if a.numel() != 0 or n.numel() != 0:
+                    if not torch.allclose(a, n, rtol, atol):
+                        return fail_test('Jacobian mismatch for output %d with respect to input %d,\n'
+                                         'numerical:%s\nforward analytical:%s\n' % (i, j, n, a))
```
nit: continues can reduce the indentation, but it probably comes down to personal preference.
```
if a.numel() == 0 and n.numel() == 0:
    continue
if torch.allclose(a, n, rtol, atol):
    continue
return fail_test(...)
```

zou3519 replied:
Also, if a.numel() == 0 and n.numel() == 0, shouldn't we check that the shapes of the tensors are the same?

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -545,6 +613,7 @@ def gradgradcheck(
             the second derivative.
         check_undefined_grad (bool, options): if True, check if undefined output grads
             are supported and treated as zeros
+        check_forward(bool, options): if True, check the forward mode AD gradient
```
options -> optional

albanD replied:
Copy pasted the options from the one above... But I think you fixed this one in master? I'll update!

zou3519 replied:
My PR got reverted haha. I'll re-submit it later today

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -278,6 +315,8 @@ def gradcheck(
             exactly (default, 0.0) or be within this tolerance.
         check_undefined_grad (bool, options): if True, check if undefined output grads
             are supported and treated as zeros, for ``Tensor`` outputs.
+        check_forward (bool, optional): if True, check the forward mode AD gradient
```
We should state the default (False)

zou3519 commented on test/test_autograd.py:
```diff
@@ -5556,7 +5559,10 @@ def bar(a):
     def test_jvp_err_check_strict(self):
         def foo(a):
-            return a.detach()
+            # break both forward and backward mode links
+            out = a.detach()
+            out, _ = fwAD.unpack_dual(out)
+            return out
```
`test_jvp_err_check_strict` doesn't call `jvp` with `fw_mode=True`, so the unpack_dual line here doesn't actually do anything meaningful, right?

On a related note, it would be good to update all of the jvp tests to test with both `fw_mode=True` and `fw_mode=False` for comprehensiveness. I see the following functions:
- test_jvp_output
- test_jvp_scalar
- test_jvp_err_check
- test_jvp_err_check_strict
- test_jvp_create_graph (has already been updated to test fw_mode=True)

zou3519 commented on test/test_autograd.py:
```diff
@@ -5654,27 +5660,45 @@ def reducer(x):
         self.assertIsNotNone(res[0].grad_fn)
         self.assertIsNotNone(res[1].grad_fn)
 
-        gradcheck(lambda inp, v: autogradF.jvp(reducer, inp, v, create_graph=True), (inputs, v))
-        gradgradcheck(lambda inp, v: autogradF.jvp(reducer, inp, v, create_graph=True), (inputs, v))
+        res = autogradF.jvp(reducer, inputs, v, create_graph=True, fw_mode=False)
+        self._assert_same_struct(res[1], res[0])
+        self.assertIsNotNone(res[0].grad_fn)
+        self.assertIsNotNone(res[1].grad_fn)
```
I'm confused -- isn't `fw_mode=False` the default right now?

Also nit:
```
for fw_mode in [True, False]:
        res = autogradF.jvp(reducer, inputs, v, create_graph=True, fw_mode=False)       
        self._assert_same_struct(res[1], res[0])
        self.assertIsNotNone(res[0].grad_fn)
        self.assertIsNotNone(res[1].grad_fn)
```

albanD replied:
It is, this is left from an older version. My bad!

zou3519 commented on test/test_autograd.py:
```diff
@@ -5654,27 +5660,45 @@ def reducer(x):
+        self.assertIsNotNone(res[0].grad_fn)
+        self.assertIsNotNone(res[1].grad_fn)
+
+        # check_forward=False here as nested forward is not supported yet
+        gradcheck(lambda inp, v: autogradF.jvp(reducer, inp, v, create_graph=True), (inputs, v), check_forward=False)
+        gradgradcheck(lambda inp, v: autogradF.jvp(reducer, inp, v, create_graph=True), (inputs, v), check_forward=False)
+        gradcheck(lambda inp, v: autogradF.jvp(reducer, inp, v, create_graph=True, fw_mode=False), (inputs, v))
+        gradgradcheck(lambda inp, v: autogradF.jvp(reducer, inp, v, create_graph=True, fw_mode=False), (inputs, v))
```
Ditto -- isn't fw_mode=False? Am I missing an override somewhere?

zou3519 reviewed (commented):
The logic in the gradcheck and jvp all looks fine to me. I have some high-level comments:

1. I would have probably separated this PR into 2 pieces: one that updates gradcheck to check forward mode, and another that turns on forward mode AD in the jvp api, but it's OK to leave it as is since as the reviewer I figured out which tests correspond to which of these changes.
2. API bikeshedding: in the APIs, `check_forward` or `fw_mode` can be ambiguous, especially when the user is familiar with terminology for "NN module forward pass". Naming it "check_forward_mode_ad" is longer but removes the ambiguity.
3. We should extend all of the jvp tests to test `fw_mode` on True and False (I think only some were updated).

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

IvanYashchuk commented on torch/autograd/gradcheck.py:
```diff
@@ -201,6 +202,40 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
 
     return jacobian, reentrant, correct_grad_sizes, correct_grad_types
 
+def get_analytical_jacobian_fw(fn, input, output):
```
To test the correctness of forward mode AD computing the whole Jacobian using finite differences and "analytical" is not needed. It's enough to compute a directional derivative for a random direction (Jacobian-vector product for a random tangent vector) 
```python
v = ... # random vector like input
analytical_vjp = pytorch_vjp(func, input, v)
numerical_vjp = finite_differences_vjp(func, input, v)
torch.allclose(analytical_vjp, numerical_vjp)
```
```python
def finite_differences_vjp(func, inp, v):
"""Jacobian of `func` at `inp` multiplied by `v`"""
    # or using any other more accurate finite difference scheme
    return (func(inp + eps * v) - func(inp)) / (eps * norm(v))
```
So approximately computing JVP by finite difference takes only two function calls. Avoiding computing the whole Jacobian would speed up testing a lot.

albanD replied:
Yes, That is also mostly true as well for backward mode.
The point here is more that having the full Jacobian is much more useful as an error message :) 

A fast version of gradcheck that uses that strategy you're talking about is in the pipe though and would also be applied to forward AD. But the full Jacobian version will still be kept to ensure good error messages in gradcheck.

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD committed:
Update on "Add gradcheck for forward AD by default and functional API"


RFC: https://github.com/pytorch/rfcs/pull/11

This PR adds the option to check forward grad using gradcheck. The current logic is:
- Forward grad is always checked
- If the forward evaluation fails because an op is not implemented, the test is silently passing

The goal is to make sure that all formulas that are added are properly tested without having to add a new test for each op.

The final logic after the next PR that adds the remaining formulas is going to be:
- Forward grad is always checked
- Failure with not implemented op is an actual failure
- Users should set `check_forward=False` if they explicitly don't want to test forward grads (which should not be the case internally).

Differential Revision: [D25607502](https://our.internmc.facebook.com/intern/diff/D25607502)

[ghstack-poisoned]

albanD commented:
This is being replaced by https://github.com/pytorch/pytorch/pull/57633
